### ISSUE TITLE
When should I close Cache?

### ISSUE THREAD

**Issue by Sraw** [grantjenks/python-diskcache#71]
The tutorial is ambigous, and I don't understand clearly when should I close Cache.

Should I close it during every batch operations? Or should I close it before application exits?

And after close, couldn't I reopen connection based on exsiting instance? Do I have to create another instance with the same parameters?

**Comment by grantjenks** [comment]
Open and close it as little as possible. Every operation is atomic.

With the database open, it uses one or two file handles and caches keys/values in memory. The memory usage is configurable using disk cache settings.

Pull requests to improve the documentation are welcome.

**Comment by abubelinha** [comment]
Hi. Thanks @grantjenks for this really useful package.

Related to this closed issue, I have some newbie questions (sorry if I am too verbose):

**(1) Are open/close operations always needed?**

I ask it because of this [tutorial](https://grantjenks.com/docs/diskcache/tutorial.html) part:
> _Caches have an additional feature: [memoizing](https://grantjenks.com/docs/diskcache/api.html#diskcache.FanoutCache.memoize) decorator. The decorator wraps a callable and caches arguments and return values_".

In that code example, no open/close operations are shown. Just repeated calls to the decorated function.
_______________

**(2) Does diskcache already use RAM to reduce disk accesses?**

Your [above comment](https://github.com/grantjenks/python-diskcache/issues/71#issuecomment-402593654) suggests I don't need other caching systems to reduce the number of disk-read accesses:

> With the database open, it uses one or two file handles and caches keys/values in memory
     
But I understood the opposite thing when I first read your other comment [here](https://github.com/grantjenks/python-diskcache/issues/192#issuecomment-770412879) about _layering caches_ ... like if you were suggesting us to combine diskcache with some other caching methods (i.e. `@functools.lru_cache()`) to avoid many diskcache repeated disk accesses. Perhaps you were just describing how diskcache internally works _**already**_? 
__________
**(3) Are diskcache keys auto-generated? (at least sometimes)**

I am aware you can set/get/touch ...  cache dictionary entries by key. But again, the memoizing code example in tutorial does not use them.  So, would this example code be OK (not explicitly setting any keys myself)?
```Python
from diskcache import FanoutCache
cache = FanoutCache('.mycache', statistics=True)

@cache.memoize(expire=600)
def get_from_server(sqlquery=None, mode="access", pyodbc_lowercase = True):
    # my code to produce output 
    return(output)

list_of_queries = ["SELECT ... ", "SELECT ... ", "SELECT ... ", "SELECT ... "] # all different from each other
use_cases = ["use 1", "use 2", "use 3", ... "use N"]

for i, use_case in enumerate(use_cases):
    # depending on use_case I will use same data values for different purposes:
    # (... use_case dependant code ...)
    for q in list_of_queries:
        values = get_from_server(sqlquery = q)
        # (... use_case dependant code ...)

cache.close()  # not sure if I need this or not when using memoize decorator
```
Probably it's a bad example, just to see if I understand how things change in each external for iteration. 
I believe `@cache.memoize` will intercept values passed in every call to `get_from_server()`. 
I also understand a key would be auto-generated based on the combined 3 parameter values passed.
So, in each 1-N iteration:
1. function will run, read values from server, produce output and diskcache will store it in disk, associated to the autogenerated key.
2. diskcache will recognize repeated 3-pars combined values (so repeated key) and will read function output from disk
3. diskcache will recognize repeated 3-pars combined values (so repeated key) and will read function output from RAM
4. ... and further iterations: diskcache keeps using RAM for reading function output (when values produce again the same key)

**(4)** Then, **if I re-run the script once again before the expire time, the server is never used** (1st iteration will already found the keys and values in disk, and 2-N iterations will read them from memory). Correct?

**(5) Is a memoize decorator the right approach for such an script?**  
Also, should I call `cache.close()` at its end, or perhaps decorator function acts like a `with` statement? 

Sorry if these are silly questions but my programming skills are pretty limited (first time I read about python decorators). 

Thanks a lot in advance for your answers !!

**Comment by grantjenks** [comment]
1. Nope, close() operations aren't too important. You'd only want to call these if the process ran out of file descriptors. The OS default provides hundreds or thousands of those so it'll be likely fine.
2. Yes, diskcache uses memory to cache keys/values. But that memory access still involves serialization/deserialization of keys/values. functools.lru_cache does no serialization and will be much faster. I answered https://stackoverflow.com/a/71634221/232571 with a solution for lru_cache with ttl. You can layer these if you like.
3. That's correct. cache.memoize() converts the function args/kwargs into a cache key for you.
4. Again, correct. It'll read the values from the disk. Be careful if you change the logic inside `get_from_server`. DiskCache doesn't hash the contents of a function, only it's arguments.
5. Memoize seems fine here. Beware that calling it like `get_from_server('SELECT 1', mode='access')` and `get_from_server('SELECT 1')` will appear different to memoize() even though the default argument makes the args the same.
6. Don't worry about calling cache.close(). It'll be fine.

**Comment by abubelinha** [comment]
Thank you so much for your detailed answers. Really helpful !!
:+1: :+1: :+1:

**Comment by abubelinha** [comment]
> Yes, diskcache uses memory to cache keys/values. But that memory access still involves serialization/deserialization of keys/values. functools.lru_cache does no serialization and will be much faster.

Is this speed difference related to the fact that [`@functools.lru_cache()`](https://docs.python.org/3/library/functools.html#functools.lru_cache) or similar [cachetools](https://github.com/tkem/cachetools/) cannot handle function arguments of certain "unhashable" types? (i.e. lists, dictionaries).

diskcache handles those functions without problems (I tried with lists, dicts and pandas dataframes). Your [lru_cache with ttl](https://stackoverflow.com/a/71634221/232571) does not.
Is that explained by serialization?
Perhaps it's similar to this [custom `envkey()` function](https://cachetools.readthedocs.io/en/latest/#cachetools.keys.typedkey) as a workaround to solve [this cachetools issue](https://github.com/tkem/cachetools/issues/28#issuecomment-246169713)?

Basically, I am trying to figure out if that kind of envkey() solution would make lru_cache() slower (so lru layering would not make sense for functions with unhashable arguments).

**Comment by grantjenks** [comment]
> Is that explained by serialization?

Sorta. DiskCache uses Pickle for serialization. Lists, dicts, and Pandas DataFrames are all converted to bytes for storage on disk. DiskCache compares the bytes for equality and search. It does not hash the values.

**Comment by abubelinha** [comment]
> DiskCache uses Pickle for serialization

So it's something [like this](https://github.com/tkem/cachetools/issues/97#issue-325011791)?
I am a bit confused about how did you get it to work properly, since cachetools [decided not to use](https://github.com/tkem/cachetools/issues/98) Pickle for some [reason](https://stackoverflow.com/questions/23069908/pickle-serialization-order-mystery). 
As far as I understand it was not about speed, but the uncertainty of `pickle.dumps()` always producing the same unique key.

Thanks for your patience, as we are not talking about DiskCache issues.  
Just wondering why other libraries don't use the same approach (not even as a non-default option).

**Comment by abubelinha** [comment]
@grantjenks Regarding your answer to the [3rd question above](https://github.com/grantjenks/python-diskcache/issues/71#issuecomment-1404074427):
> 3. That's correct. cache.memoize() converts the function args/kwargs into a cache key for you.

In my use case, arguments were sql queries.
"SELECT * FROM bibliography WHERE author='Grant J.';"
So, after a first use of that query string, mysql server is not called again as long as the query string does not change (or cache expires).

But suppose I add a new record to my database, where author is 'Grant J.'.
Then I know my example cached query will be obsolete. I would like to force an update of that particular cached entry.
Is there a way I can somehow force cache.memoize() to not use previous cached data, but update it and return new value in a particular call to a memoized function?

I think about two possibilities

1. get cached entry key and then delete that entry with `del cache['key']`
    But I don't know how can I first get the key of this particular cached function call?  
    `get_from_server(sqlquery = "SELECT * FROM bibliography WHERE author='Grant J.';")`

2. Or perhaps this is this another way to go?
   `cache.evict("SELECT * FROM bibliography WHERE author='Grant J.';")`

Thanks in advance

**Comment by grantjenks** [comment]
> As far as I understand it was not about speed, but the uncertainty of pickle.dumps() always producing the same unique key.

That is a valid concern. DiskCache simply accepts the limitation by default. Use JSONDisk in DiskCache for a more reliable and more limited serializer.

> But I don't know how can I first get the key of this particular cached function call?

Read the docs at https://grantjenks.com/docs/diskcache/api.html#diskcache.Cache.memoize There’s an attribute that can be used to generate the key for the args.

**Comment by abubelinha** [comment]
> Read the docs at https://grantjenks.com/docs/diskcache/api.html#diskcache.Cache.memoize There’s an attribute that can be used to generate the key for the args.

Thanks @grantjenks 
Do you mean this?

> An additional __cache_key__ attribute can be used to generate the cache key used for the given arguments.
> 
> ```
> key = fibonacci.__cache_key__(100)
> print(cache[key])
> 354224848179261915075
> ```

I assume `__cache_key__` should also work for [FanoutCache memoize](https://grantjenks.com/docs/diskcache/api.html#diskcache.FanoutCache.memoize) 
(although its code example in link still says `import Cache`, and not `import FanoutCache` ... a copy-paste mistake?)

I tried to use that in my above example code [(3)](https://github.com/grantjenks/python-diskcache/issues/71#issuecomment-1404074427), which used FanoutCache.
So, I first listed the entries in my cache (in order to see the last one):
```
list(cache)

[('__main__.get_from_server', None, 'db', 'C:\\bib.mdb', 'mode', 'access', 'sqlquery', 'SELECT * from bib where record in (19, 20)'), 
('__main__.get_from_server', None, 'db', 'C:\\bib.mdb', 'mode', 'access', 'sqlquery', 'SELECT * from bib where record in (149)'), 
('__main__.get_from_server', None, 'db', 'C:\\bib.mdb', 'mode', 'access', 'sqlquery', 'SELECT * from bib where record in (86, 52)'), 
('__main__.get_from_server', None, 'db', 'C:\\bib.mdb', 'mode', 'access', 'sqlquery', 'SELECT * from bib where record in (89)')]
```
So my function `get_from_server(db, mode, sqlquery)` takes three parameters.
Now, figure out I modify my database entry with record number 89, so I'd like to update the last item in cache.

Looking to the fibonacci example calling `function.__cache_key__(value)` without parameter names, I understand this the syntax I should use for my 3-parameter function:
```
k = get_from_server.__cache_key__('C:\\bib.mdb', 'access', 'SELECT * from bib where record in (89)')
print(k) # just to see the key
print(cache[k])
```
OUTPUT:
```
('__main__.get_from_server', 'C:\\bib.mdb', 'access', 'SELECT * from bib where record in (89)', None)
Traceback (most recent call last):
  File "cache_test.py", line 1210, in <module>
    print(cache[k])
  File "C:\Python38\lib\site-packages\diskcache\fanout.py", line 300, in __getitem__
    return shard[key]
  File "C:\Python38\lib\site-packages\diskcache\core.py", line 1234, in __getitem__
    raise KeyError(key)
KeyError: ('__main__.get_from_server', 'C:\\bib.mdb', 'access', 'SELECT * from bib where record in (89)', None)
```
Looks like I am not 
What am I doing wrong?

What about the other way I proposed, using `evict('SELECT * from bib where record in (89)')`?
Should that delete my last entry too?

**Comment by abubelinha** [comment]
I am wondering about two possibilities:
1. What if my function has other named parameters with default values which I usually don't pass when calling it?
2. As all they are named parameters, what if I passed them in a different order when I called the function?

I simplified my example code (3) above. In my real function, definition is like this (5 parameters):
`get_from_server(db=None, sqlquery=None, mode="access", debug=False, pyodbc_lowercase = True)`

In my test script I only passed the first 3 parameters.

But when listing the cache:
- `debug` and `pyodbc_lowercase` are not shown 
- `sqlquery` was the 2nd parameter, but it is shown 3rd position of listed cache entries
